aten/src/ATen/native/FractionalMaxPool2d.cpp

```diff
@@ -0,0 +1,363 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+#include <tuple>
+#include <vector>
+
+namespace at {
+namespace native {
+namespace {
+
+template <typename scalar_t>
+static std::vector<int> fractional_max_pool2d_generate_intervals(
+  scalar_t sample,
+  int inputSize,
+  int outputSize,
+  int poolSize) {
+  scalar_t alpha = (scalar_t) (inputSize - poolSize) / (scalar_t) (outputSize - 1);
+  std::vector<int> sequence(outputSize, 0);
+
+  for (int i = 0; i < outputSize - 1; ++i) {
+    sequence[i] =
+      (int) ((i + sample) * alpha) - (int) (sample * alpha);
+  }
+  sequence[outputSize - 1] = inputSize - poolSize;
+
+  return sequence;
+}
+
+template <typename scalar_t>
+static void fractional_max_pool2d_out_frame(
+  scalar_t* input,
+  scalar_t* output,
+  int64_t* indices,
+  scalar_t* randomSamples,
+  int numPlanes,
+  int inputW, int inputH,
+  int outputW, int outputH,
+  int poolSizeW, int poolSizeH) {
+  int plane;
+#pragma omp parallel for private(plane)
+  for (plane = 0; plane < numPlanes; ++plane) {
+    /* each plane contains 2 random samples, one for W and one for H */
+    scalar_t* randomSamplesForPlane = randomSamples + plane * 2;
+
+    /* Generate interval sequence */
+    auto sequenceW = fractional_max_pool2d_generate_intervals<scalar_t>(
+        randomSamplesForPlane[0], inputW, outputW, poolSizeW);
+    auto sequenceH = fractional_max_pool2d_generate_intervals<scalar_t>(
+        randomSamplesForPlane[1], inputH, outputH, poolSizeH);
+
+    /* loop over output */
+    int h, w;
+
+    scalar_t* inputForPlane = input + plane * inputW * inputH;
+    scalar_t* outputForPlane = output + plane * outputW * outputH;
+    int64_t* indicesForPlane = indices + plane * outputW * outputH;
+
+    for (h = 0; h < outputH; ++h) {
+      int inputHStart = sequenceH[h];
+
+      for (w = 0; w < outputW; ++w) {
+        int inputWStart = sequenceW[w];
+
+        scalar_t maxVal = -std::numeric_limits<scalar_t>::infinity();
+        int64_t maxIndex = -1;
+
+        int h2, w2;
+        for (h2 = inputHStart; h2 < inputHStart + poolSizeH; ++h2) {
+          for (w2 = inputWStart; w2 < inputWStart + poolSizeW; ++w2) {
+            AT_ASSERT(h2 >= 0 && h2 < inputH);
+            AT_ASSERT(w2 >= 0 && w2 < inputW);
+
+            int planeIndex = h2 * inputW + w2;
+            scalar_t val = inputForPlane[planeIndex];
+            if (val > maxVal) {
+              maxVal = val;
+              maxIndex = planeIndex;
+            }
+          }
+        }
+
+        AT_ASSERT(maxVal != -std::numeric_limits<scalar_t>::infinity());
+        AT_ASSERT(maxIndex != -1);
+
+        outputForPlane[h * outputW + w] = maxVal;
+        /* +1 to lua index */
+        indicesForPlane[h * outputW + w] = maxIndex;
+      }
+    }
+
+  }
+}
+
+void fractional_max_pool2d_out_cpu_template(
+  at::Tensor const& input_,
```

Comment:
nit: In this codebase, we use the more conventional `const at::Tensor&` spelling. Best not to force people to have to know how both spellings work :)